### ISSUE TITLE
How to select multiple attributes efficiently with QueryBuilder?

### ISSUE THREAD

**Issue by nicoladonadoni** [typeorm/typeorm#5377]
**Issue type:**

[X] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[X] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I have a query that operates on 6 tables and returns some attributes from each of them, for a total of over 100 attributes selected. I have noticed that if I only return 1 attribute per table (i.e. just the "id") on my machine the code is 2.5 seconds faster than if I select all the 100+ attributes.

If my query is a getMany() which joins as N:M another table, the number of records in the table M doesn't affect the time that the code takes to return me the data but if I remove the selection of attributes (~50 of them) from such table M, then the code runs much faster (even if the table itself has 0 records in it!).

Please note that this is not a query optimisation question but it's related to how TypeORM reconstructs the final object returned from a getOne() or getMany() function.

Any help will be greatly appreciated, thanks!

**Comment by boenrobot** [comment]
I would suggest you first setup some monitoring/proxy, that will let you capture the exact queries that TypeORM generates.

After you have the different queries, you can do an EXPLAIN on both to figure out what about it slows it down.

I'm guessing that one of the columns you end up with is also used in condition (be it a join, a where or having) AND is not indexed. If that's the case, adding the appropriate index will likely help. But that's a very blind guess.

**Comment by nicoladonadoni** [comment]
Thank you for your answer but I think you didn't read properly: I mentioned this is not a query optimization problem (the query runs quickly). I have started to debug the TypeORM library and it seems to be slow when composing the query, so when it has to transform something like 

```
.select(['Table1.column1', 'Table1.column2', '...', 'Table1.columnN'])
.addSelect(['Table2'.column1', 'Table2'.column2, '...', 'Table2'.columnM'])
.from('Table1')
.innerJoinAndMapMany(...)
```

to the actual SQL query and it's dependent to the number of columns that are included in the .select or .addSelect functions

**Comment by boenrobot** [comment]
I did read that part, but it's not clear to me whether your measurements are factoring in the execution of the query in both cases, so you should make sure to know the difference there, and remove it from the total difference.

Or, to make sure you're measuring the effect of TypeORM's query builder only, you can use getSql() instead of getMany()/getOne(), since the later just executes the result of getSql().

**Comment by nicoladonadoni** [comment]
Sorry I had deleted my answer because I was digging more into the code. Your suggestion was great and actually I have done what you said and this is the result:

I have the following sample code:
```
const results = await Promise.all([
    runQuery1(),
    runQuery2(),
    runQuery3()
])
```
where each of the runQuery functions just does the following
```
const sqlQuery = typeorm_1.getConnection().createQueryBuilder()
            .select(['Table1.column1', 'Table1.column2', ... 'Table1.columnN'])
            .addSelect(['Table2.column1', 'Table2.column2', ... 'Table2.columnM'])
            .from('Table1')
            .innerJoinAndMapMany('...')
            .where('....')
            .orderBy('...')
            .setParameters({ ... })
            .getSql();
return sqlQuery;
```

This executes in ~3s while if I remove all the selected columns and just keep column1 for all the tables for example (so basically I reduce the number of columns to select drastically from 100+ to just a few then the code executes in a few milliseconds.

It's scary because if I add other queries in parallel and have a total of 1000+ selected columns the code takes like more than 10 seconds, and it's only generating the SQL query, not executing it! 

I hope this gives you more insights, I really look forward to understanding why this is happening, thanks!

**Comment by boenrobot** [comment]
I haven't delved into the source, but AFAIK, the query builder allows you to use entity names and properties. So what I think is happening is that for each column, TypeORM is trying to find a matching property among all known entities, and get the SQL column name out of it. Even if you're using the SQL names, the query builder is still doing this lookup before giving up, and falling back to using the names as given.

I'm not aware of a way to disable this feature, and thus the lookup.

**Comment by nicoladonadoni** [comment]
Well, actually it has to do the lookup because I don't use the real PostgreSQL column name but the entity class attribute's name, so the lookup is needed.

The biggest problem here is that it's making the event loop lag! So for those N seconds that it takes to do the mapping and generate the SQL query, my server can't handle other requests or do nothing at all.

I wonder if I should report this as a bug? It seems pretty serious in my opinion if someone else can confirm it.

**Comment by boenrobot** [comment]
You could store the generated query outside of the request, and just run it with parameters from the request.

**Comment by OliverLeighC** [comment]
I've been running into this issue as well, the slowness in mapping the query result to the entities. 

It seems to be this chunk of code here that is especially slow https://github.com/typeorm/typeorm/blob/fc7a4f13f6270c9d5206358ebd78adb0ca690b3d/src/query-builder/SelectQueryBuilder.ts#L1911 

When we use getRawMany() it's significantly faster, but getMany() which does map to our entities is causing performance issues.

**Comment by imnotjames** [comment]
Duplicate of #2381
